### PR TITLE
[JENKINS-48840] - PCT should fallback to a default GitHub location when checkout by scm/connection fails

### PR DESCRIPTION
**Pull request by oleg-nenashev**
https://issues.jenkins-ci.org/browse/JENKINS-48840

when you runs a PCT without specifying local checkout sources, the PCT takes the source location from a "scm/connection" field of the POM in the plugin info. In such case PCT will fail if...

* Connection info has a mistake (e.g. Custom Tools Plugin). https://github.com/jenkinsci/customtools-plugin/blob/custom-tools-plugin-0.5/pom.xml#L44 
* Connection info points to SSH instead of HTTPs. It will pass iff you have a default key in the environment available w/o passwords. E.g. https://github.com/jenkinsci/coverity-plugin/blob/coverity-1.10.0/pom.xml#L83

Probably the same patch needs to be applied to multi-module repos, will confirm it later

@reviewbybees

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Test failures should be ignored while #59 is pending. The tests are green here though

**Comment by recena** [comment]
In first case, the SCM info would be fixed. No?

**Comment by recena** [comment]
In the second case, the repositories should be updated to use HTTPS as GitHub recommends 😄

**Comment by oleg-nenashev** [comment]
> In the second case, the repositories should be updated to use HTTPS as GitHub recommends 😄

I am happy to let you to convince all the maintainers :trollface:

**Comment by jglick** [comment]
> Connection info has a mistake

Then just fix it, and release the fix. PCT should not be made to work around this.

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 1]
```diff
@@ -411,13 +412,21 @@ private TestExecutionResult testPluginAgainst(MavenCoordinates coreCoordinates,
...
-                    CheckOutScmResult result = scmManager.checkOut(repository, new ScmFileSet(pluginCheckoutDir), new ScmTag(plugin.name + "-" + plugin.version));
+                    String tag = plugin.name + "-" + plugin.version;
+                    CheckOutScmResult result = tryPluginSrcCheckoutFromScm(pomData.getConnectionUrl(), pluginCheckoutDir, tag);
 
                     if (!result.isSuccess()) {
-                        throw new RuntimeException(result.getProviderMessage() + " || " + result.getCommandOutput());
+                        StringBuilder fallbackCheckoutUrl = new StringBuilder("scm:git:git://github.com/jenkinsci/");
+                        fallbackCheckoutUrl.append(pomData.artifactId);
+                        if (!pomData.artifactId.endsWith("-plugin")) {
+                            fallbackCheckoutUrl.append("-plugin");
```
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: Will not work very generally.

**Comment by oleg-nenashev** [comment]
@jglick 

> This is the wrong fix. See ScmConnectionSpecialCasesTest and PluginRemoting.computeScmConnection.

Could you clarify? Do you propose to add SSH resolution to `luginRemoting.computeScmConnection` or what?

**Comment by raul-arabaolaza** [comment]
I agree with @jglick in that the PCT should not try to workaround errors in connection url, that way you are just hiding a bug. 

About the ssh thing, seems to me like when you need to update the parent pom because the existing one is too old, that is done mainly in [MavenPom](https://github.com/jenkinsci/plugin-compat-tester/blob/2930c8df6bf5ac3ea79a07eacc33480331997d26/plugins-compat-tester/src/main/java/org/jenkins/tools/test/model/MavenPom.java) and [TransformPom](https://github.com/jenkinsci/plugin-compat-tester/blob/2930c8df6bf5ac3ea79a07eacc33480331997d26/plugins-compat-tester/src/main/java/org/jenkins/tools/test/hook/TransformPom.java). So maybe using that hook to change the ssh url to the recommended one could be a good idea. As in this case there is nothing "wrong" I believe is interesting for the PCT to be able to workaround the situation

**Comment by oleg-nenashev** [comment]
Waiting for the response from @jglick about the preferred approach for SSH destinations

**Comment by recena** [comment]
👎 with this approach.

**Comment by jglick** [comment]
also needs merge conflict resolution

**Comment by jglick** [comment]
Abandoned and presumed obsolete.

**Pull request closed without merging**